### PR TITLE
Add decompositions for avg_poolnd

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116418
* #116491

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Comment by isuruf** [comment]
There is a lowering in `torch/_inductor/lowering.py` that works for a total window size `< 25`. Benchmarking against that, we get,

```
[--------------------------- avg_pool --------------------------]
                             |  Decomposed  |  Lowering  |  Eager
32 threads: -----------------------------------------------------
      (3, 5, 1000000)        |     303      |    303     |   455 
      (3, 5, 1000, 1000)     |     250      |    250     |   310 
      (3, 5, 100, 100, 100)  |     230      |    272     |   274
```
for window size equal to `2` per dimension. The generated code for 1D, 2D are exact.
 
For window size equal to `30` per dimension, the lowering falls back to the eager op, which leads to,

```
[--------------------------- avg_pool --------------------------]
                             |  Decomposed  |  Lowering  |  Eager
32 threads: -----------------------------------------------------
      (3, 5, 1000000)        |     207      |     630    |    633
      (3, 5, 1000, 1000)     |     190      |     611    |    614
      (3, 5, 100, 100, 100)  |     156      |    1682    |   1684
```

Benchmark script

<details>

```python
import torch
from torch.testing import make_tensor
from torch.fx.experimental.proxy_tensor import make_fx
from torch.utils.benchmark import Timer, Compare
from torch._inductor.compile_fx import compile_fx_inner, cudagraphify_impl
from torch._inductor.decomposition import decompositions
from itertools import product
from functools import partial
aten = torch.ops.aten

torch._logging.set_logs(output_code=True)

benchmark_name = "avg_pool"
Ss = [[3, 5, 1000000], [3, 5, 1000, 1000], [3, 5, 100, 100, 100]]


def gen_inputs():
    for shape in Ss:
        yield [torch.randn(shape, dtype=torch.float32, device="cuda")]

def benchmark(label, f, x):
    return Timer("f([x,])",
                 globals=locals(),
                 label=benchmark_name,
                 description=label,
                 sub_label=f"{tuple(x.shape)}",
                 num_threads=torch.get_num_threads()).blocked_autorange(min_run_time=2)

def compare(args):
    x, = args
    print(f"{tuple(x.shape)}")
    if x.dim() == 3:
        aten_func = aten.avg_pool1d
    elif x.dim() == 4:
        aten_func = aten.avg_pool2d
    elif x.dim() == 5:
        aten_func = aten.avg_pool3d
    
    def f(args):
        x, = args
        val = aten_func(x, 30)
        return (val,)

    decomposed = make_fx(f, decomposition_table=decompositions, tracing_mode="fake")(args)
    compiled_decomposed = compile_fx_inner(decomposed, args, cudagraphs=False)
    yield benchmark(f"Decomposed", compiled_decomposed, *args)

    non_decomposed = make_fx(f, tracing_mode="fake")(args)
    compiled_nondecomposed = compile_fx_inner(non_decomposed, args, cudagraphs=False)
    yield benchmark("Lowering", compiled_nondecomposed, *args)
    # Just show the first two generated kernels
    #torch._logging.set_logs(output_code=False)

    cuda_f = cudagraphify_impl(f, args, static_input_idxs=tuple(range(len(args))))
    yield benchmark(f"Eager", cuda_f, *args)

results = []
for x in gen_inputs():
    for res in compare(x):
        results.append(res)

compare = Compare(results)
compare.trim_significant_figures()
compare.print()
```

</details>

**Review by lezcano — COMMENTED** [review]
The results of the numbers for 1d and 2d of the decomposition vs lowering are exactly the same. Could it be the case that you are not testing the decomposition in that case, but just the lowering?

I would assume that the lowering should be faster here, given that it uses masked operations, which we don't have in decompositions.

As a side note, we should consider implementing a decomposition-only op that lowers into masked. I think it would be enough for our use case to have an advanced indexing operation, together with a mask of the same shape as the index that checks whether the input should be loaded or not. An even better way of doing this would be to detect the pattern advanced indexing -> clamp -> where and convert it into a masked load.

**Comment by lezcano** [comment]
Actually, I don't think we have a nice way to perform graph transformations at an IR level... We could do something like what we do for the int64 -> int32 indexing transformation, but we would need to implement a pattern matcher at an IR level...

@jansel what would be the best to implement the graph transformation described at https://github.com/pytorch/pytorch/pull/116418#pullrequestreview-1796908802?

**Comment by isuruf** [comment]
> The results of the numbers for 1d and 2d of the decomposition vs lowering are exactly the same. Could it be the case that you are not testing the decomposition in that case, but just the lowering?

No, it's because they generate the exact same triton code. I double checked by removing the lowering.

> I would assume that the lowering should be faster here, given that it uses masked operations, which we don't have in decompositions.

I'm using `torch.where` here which is limited than a masked operation, but is enough for our implementation here.

> As a side note, we should consider implementing a decomposition-only op that lowers into masked. I think it would be enough for our use case to have an advanced indexing operation, together with a mask of the same shape as the index that checks whether the input should be loaded or not. An even better way of doing this would be to detect the pattern advanced indexing -> clamp -> where and convert it into a masked load.

Agreed. The workaround using `torch.where` here is a bit hard to read, and it would be good to have a new op.

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -3256,16 +3256,14 @@ def fn(x):
...
+        # Large kernel size
         def fn(x):
             return aten.avg_pool2d(x, [13, 13], [1, 1], [0, 0])
 
-        torch._inductor.metrics.generated_kernel_count = 0
         self.common(
             fn,
             (-torch.arange(1 * 24 * 24, dtype=torch.float32).view(1, 1, 24, 24),),
         )
-        self.assertEqual(torch._inductor.metrics.generated_kernel_count, 0)
```
[the remark this answers is not in the document]
Comment (isuruf) [reply test/inductor/test_torchinductor.py:3268]: This was checking no inductor kernel was generated for large kernel sizes (window size >25) since the lowering fallbacks to eager mode for that case. The decomposition works for any window size and therefore this test is wrong now.

[the remark this answers is not in the document]
Comment (lezcano) [reply test/inductor/test_torchinductor.py:3268]: Note that the fallback was there not for correctness but for performance. We should find what are relevant shapes to test this operation on and make sure that we are not regressing on those.

**Review by jansel — CHANGES_REQUESTED** [review]
Please add some benchmarks for 3x3 and 9x9, which are both common. 
 You should also use real sizes for images taken from our benchmarks.  TIMM benchmarks should contain avg_pool2d ops.

The masking should mainly matter for smaller image sizes where the cost of computing the boundary is a larger fraction of the total compute.

**Comment by isuruf** [comment]
Thanks @jansel for the review. I looked at input sizes from `timm_resnet` benchmark and the decomposition is slower than the lowering. Comparing the generated codes from the decomposition and the lowering, the difference seems to come from the thresholds,
```
realize_reads_threshold
realize_bytes_threshold
unroll_reductions_threshold
```
in `torch._inductor.config`. Is there a way to control these for a decomposition or should I manually do the unrolling in the decomposition.

**Review by lezcano — COMMENTED** [review]
Do not duplicate the code. Add a flag to choose between unrolling or not.

**Comment by jansel** [comment]
> in torch._inductor.config. Is there a way to control these for a decomposition or should I manually do the unrolling in the decomposition?

We should try not to special case things like this.  This likely points to something suboptimal going on in our realize/unroll heuristics.  I'd suggest trying to update the heuristics to they produce better results.  Improved heuristics may also help other kernels.

File: torch/_inductor/decomposition.py [hunk 2]
```diff
@@ -648,3 +655,40 @@ def masked_scatter(self, mask, source):
...
+    # TODO: remove this when #100331 is merged. We only do this
+    # for window_size <=25 to avoid performance regressions compared
+    # to the previous algorithm which unrolled manually for <=25
+    context = (
+        config.patch(unroll_reductions_threshold=25)
+        if window_size <= 25
+        else contextlib.nullcontext()
+    )
+
+    with context:
```
Comment (lezcano) [inline torch/_inductor/decomposition.py:683]: This is way too hacky. A more pythonic way would be:
```python
if window_size <= 25:
  # manually unroll via a for-loop
else:
  # x.sum()
```
In any case, you should discuss these decomps with @peterbell10 as we may need to think a bit about how we implement the decomp itself to make it fast in an A100. We had some ideas in a discussion today that may be worth trying.

**Comment by lezcano** [comment]
@isuruf what happened with this PR?

**Comment by isuruf** [comment]
The decomposition uses the same algorithm as https://github.com/pytorch/pytorch/pull/116756 and the performance is really bad.
